File: conductor/src/main/java/com/bluelinelabs/conductor/internal/StringSparseArrayParceler.java
package com.bluelinelabs.conductor.internal;

import android.os.Parcel;
import android.os.Parcelable;
import android.util.SparseArray;

public class StringSparseArrayParceler implements Parcelable {

    private final SparseArray<String> stringSparseArray;

    public StringSparseArrayParceler(SparseArray<String> stringSparseArray) {
        this.stringSparseArray = stringSparseArray;
    }

    private StringSparseArrayParceler(Parcel in) {
        stringSparseArray = new SparseArray<>();

        final int size = in.readInt();

        for (int i = 0; i < size; i++) {
            stringSparseArray.put(in.readInt(), in.readString());
        }
    }

    public SparseArray<String> getStringSparseArray() {
        return stringSparseArray;
    }

    public int describeContents() {
        return 0;
    }

    public void writeToParcel(Parcel out, int flags) {
        final int size = stringSparseArray.size();

        out.writeInt(size);

        for (int i = 0; i < size; i++) {
            int key = stringSparseArray.keyAt(i);

            out.writeInt(key);
            out.writeString(stringSparseArray.get(key));
        }
    }

    public static final Parcelable.Creator<StringSparseArrayParceler> CREATOR = new Parcelable.Creator<StringSparseArrayParceler>() {
        public StringSparseArrayParceler createFromParcel(Parcel in) {
            return new StringSparseArrayParceler(in);
        }

        public StringSparseArrayParceler[] newArray(int size) {
            return new StringSparseArrayParceler[size];
        }
    };

}

File: conductor/src/main/java/com/bluelinelabs/conductor/internal/StringSparseArrayParceler.kt
package com.bluelinelabs.conductor.internal

import android.os.Parcel
import android.os.Parcelable
import android.util.SparseArray

internal class StringSparseArrayParceler(val stringSparseArray: SparseArray<String>) : Parcelable {

  override fun describeContents(): Int = 0

  override fun writeToParcel(out: Parcel, flags: Int) {
    val size = stringSparseArray.size()
    out.writeInt(size)
    for (i in 0 until size) {
      val key = stringSparseArray.keyAt(i)
      out.writeInt(key)
      out.writeString(stringSparseArray[key])
    }
  }

  companion object {

    @Suppress("unused")
    @JvmField
    val CREATOR: Parcelable.Creator<StringSparseArrayParceler> =
      object : Parcelable.Creator<StringSparseArrayParceler> {
        override fun createFromParcel(parcel: Parcel): StringSparseArrayParceler {
          val stringSparseArray = SparseArray<String>()
          val size = parcel.readInt()
          for (i in 0 until size) {
            stringSparseArray.put(parcel.readInt(), parcel.readString())
          }
          return StringSparseArrayParceler(stringSparseArray)
        }

        override fun newArray(size: Int): Array<StringSparseArrayParceler?> = arrayOfNulls(size)
      }
  }
}